### PR TITLE
ES Module update

### PR DESCRIPTION
**Pull request by haydenyoung**
> ⚠️  NOTE: Please don't use the PR description for communication, use comments instead.

## Description (Required)

Convert ipfs-log to ES6-compatible module for more modern usage. 

## Other changes (e.g. bug fixes, UI tweaks, refactors)

- [x] Add unit test
- [x] ES update
- [x] Update documentation
- [x] Dependency upgrades

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,12 +22,12 @@
   "dependencies": {
     "json-stringify-deterministic": "^1.0.1",
     "multihashing-async": "^2.0.1",
-    "orbit-db-identity-provider": "^0.3.1",
-    "orbit-db-io": "^1.1.2",
-    "p-do-whilst": "^1.1.0",
-    "p-each-series": "^2.1.0",
-    "p-map": "^4.0.0",
-    "p-whilst": "^2.1.0"
+    "orbit-db-identity-provider": "^0.4.1-699a573.0",
```
Comment (haadcode) [inline package.json:25]: A side note and not required to change for the PR, but we can use the "next" tag here (`"orbit-db-identity-provider": "next"`) which will then pull the _latest_ version that is tagged "next" in npm. This way we don't need update all the package.json files if/when we update something in the other packages. Sorry I didn't notice to mention this in the earlier PRs.

File: test/log-append.spec.js [hunk 2]
```diff
@@ -1,19 +1,15 @@
...
-const {
-  config,
-  testAPIs,
-  startIpfs,
-  stopIpfs
-} = require('orbit-db-test-utils')
+import { config, testAPIs, startIpfs, stopIpfs } from 'orbit-db-test-utils'
+
+const { createIdentity } = IdentityProvider
+const { sync } = rimraf
```
Comment (haadcode) [inline test/log-append.spec.js:12]: Same comment here as above where rimraf is used.

File: test/log-crdt.spec.js [hunk 3]
```diff
@@ -1,19 +1,15 @@
...
 // Test utils
-const {
-  config,
-  testAPIs,
-  startIpfs,
-  stopIpfs
-} = require('orbit-db-test-utils')
+import { config, testAPIs, startIpfs, stopIpfs } from 'orbit-db-test-utils'
+
+const { sync } = rimraf
```
Comment (haadcode) [inline test/log-crdt.spec.js:11]: Same comment here as above where rimraf is used.

File: test/log-heads-tails.spec.js [hunk 4]
```diff
@@ -1,20 +1,18 @@
...
 // Test utils
-const {
-  config,
-  testAPIs,
-  startIpfs,
-  stopIpfs
-} = require('orbit-db-test-utils')
+import { config, testAPIs, startIpfs, stopIpfs } from 'orbit-db-test-utils'
+
+const { sync } = rimraf
```
Comment (haadcode) [inline test/log-heads-tails.spec.js:12]: Same comment here as above where rimraf is used.

File: test/log-iterator.spec.js [hunk 5]
```diff
@@ -1,20 +1,17 @@
...
 // Test utils
-const {
-  config,
-  testAPIs,
-  startIpfs,
-  stopIpfs
-} = require('orbit-db-test-utils')
+import { config, testAPIs, startIpfs, stopIpfs } from 'orbit-db-test-utils'
+
+const { sync } = rimraf
```
Comment (haadcode) [inline test/log-iterator.spec.js:12]: Same comment here as above where rimraf is used.

File: test/log-join-concurrent.spec.js [hunk 6]
```diff
@@ -1,18 +1,15 @@
...
 // Test utils
-const {
-  config,
-  testAPIs,
-  startIpfs,
-  stopIpfs
-} = require('orbit-db-test-utils')
+import { config, testAPIs, startIpfs, stopIpfs } from 'orbit-db-test-utils'
+
+const { sync } = rimraf
```
Comment (haadcode) [inline test/log-join-concurrent.spec.js:10]: Same comment here as above where rimraf is used.

File: test/log-join.spec.js [hunk 7]
```diff
@@ -1,21 +1,19 @@
...
-  config,
-  testAPIs,
-  startIpfs,
-  stopIpfs
-} = require('orbit-db-test-utils')
+import { config, testAPIs, startIpfs, stopIpfs } from 'orbit-db-test-utils'
+
+const { fromEntry } = Log
+const { create } = Entry
+const { sync } = rimraf
```
Comment (haadcode) [inline test/log-join.spec.js:15]: Same comment here as above where rimraf is used.

File: test/log-load.spec.js [hunk 8]
```diff
@@ -1,32 +1,32 @@
...
+import Keystore from 'orbit-db-keystore'
+import LogCreator from './utils/log-creator.js'
+import { hello, helloWorld, helloAgain } from './fixtures/v0-entries.fixture.js'
+import { v1Entries } from './fixtures/v1-entries.fixture.js'
+
+// Test utils
+import { config, MemStore, testAPIs, startIpfs, stopIpfs } from 'orbit-db-test-utils'
+
+const { length: _length } = v1Entries
+const { sync } = rimraf
```
Comment (haadcode) [inline test/log-load.spec.js:19]: Same comment here as above where rimraf is used.

File: test/log-load.spec.js [hunk 9]
```diff
@@ -837,105 +837,105 @@ Object.keys(testAPIs).forEach((IPFS) => {
...
-      const entries = [v0Entries.hello, v0Entries.helloWorld, v0Entries.helloAgain]
+      const entries = [hello, helloWorld, helloAgain]
       before(async () => {
-        await Promise.all(entries.map(e => io.write(ipfs, Entry.getWriteFormat(e), Entry.toEntry(e), { links: Entry.IPLD_LINKS })))
+        await Promise.all(entries.map(e => io.write(ipfs, getWriteFormat(e), toEntry(e), { links: IPLD_LINKS })))
       })
 
       it('creates a log from v0 json', async () => {
-        const headHash = await io.write(ipfs, 'dag-pb', Entry.toEntry(v0Entries.helloAgain), { links: Entry.IPLD_LINKS })
+        const headHash = await io.write(ipfs, 'dag-pb', toEntry(helloAgain), { links: IPLD_LINKS })
```
Comment (haadcode) [inline test/log-load.spec.js:881]: Could prolly do the same here as for all other functions for consistency, ie. `await write(...)`. Applies to all other places where io.write/read are used.

File: test/log-references.spec.js [hunk 10]
```diff
@@ -1,19 +1,15 @@
...
 // Test utils
-const {
-  config,
-  testAPIs,
-  startIpfs,
-  stopIpfs
-} = require('orbit-db-test-utils')
+import { config, testAPIs, startIpfs, stopIpfs } from 'orbit-db-test-utils'
+
+const { sync } = rimraf
```
Comment (haadcode) [inline test/log-references.spec.js:11]: Same comment here as above where rimraf is used.

File: test/log-references.spec.js [hunk 11]
```diff
@@ -72,14 +68,14 @@ Object.keys(testAPIs).forEach((IPFS) => {
...
-        assert.strict.equal(log3.values[log3.length - 1].refs.length, 3)
-        assert.strict.equal(log4.values[log4.length - 1].refs.length, 4)
+        strict.equal(log1.values[log1.length - 1].next.length, 1)
+        strict.equal(log2.values[log2.length - 1].next.length, 1)
+        strict.equal(log3.values[log3.length - 1].next.length, 1)
+        strict.equal(log4.values[log4.length - 1].next.length, 1)
+        strict.equal(log1.values[log1.length - 1].refs.length, 1)
+        strict.equal(log2.values[log2.length - 1].refs.length, 2)
+        strict.equal(log3.values[log3.length - 1].refs.length, 3)
+        strict.equal(log4.values[log4.length - 1].refs.length, 4)
```
Comment (haadcode) [inline test/log-references.spec.js:78]: Are these on purpose `strict.equal` whereas elsewhere they're `strictEqual`?

File: test/log.spec.js [hunk 12]
```diff
@@ -1,28 +1,27 @@
...
-  config,
-  testAPIs,
-  startIpfs,
-  stopIpfs
-} = require('orbit-db-test-utils')
+import { config, testAPIs, startIpfs, stopIpfs } from 'orbit-db-test-utils'
+const { LastWriteWins } = LogSorting
+const FirstWriteWins = (a, b) => LastWriteWins(a, b) * -1
+
+const { sync } = rimraf
```
Comment (haadcode) [inline test/log.spec.js:21]: Same comment here as above where rimraf is used.

File: test/replicate.spec.js [hunk 13]
```diff
@@ -1,22 +1,16 @@
...
-  testAPIs,
-  startIpfs,
-  stopIpfs,
-  getIpfsPeerId,
-  waitForPeers,
-  connectPeers
-} = require('orbit-db-test-utils')
+import { config, testAPIs, startIpfs, stopIpfs, getIpfsPeerId, waitForPeers, connectPeers } from 'orbit-db-test-utils'
+
+const { sync } = rimraf
```
Comment (haadcode) [inline test/replicate.spec.js:11]: Same comment here as above where rimraf is used.

File: test/signed-log.spec.js [hunk 14]
```diff
@@ -1,19 +1,15 @@
...
 // Test utils
-const {
-  config,
-  testAPIs,
-  startIpfs,
-  stopIpfs
-} = require('orbit-db-test-utils')
+import { config, testAPIs, startIpfs, stopIpfs } from 'orbit-db-test-utils'
+
+const { sync } = rimraf
```
Comment (haadcode) [inline test/signed-log.spec.js:11]: Same comment here as above where rimraf is used.

**Review by haadcode — CHANGES_REQUESTED** [review]
Looking good! 👌 

Few naming things to change, couple of questions.

File: test/browser.spec.js [hunk 15]
```diff
@@ -1,29 +1,29 @@
...
+  const keyF = import('./fixtures/keys/identity-keys/032f7b6ef0432b572b45fcaf27e7f6757cd4123ff5c5266365bec82129b8c5f214.json')
+  const keyG = import('./fixtures/keys/identity-keys/02a38336e3a47f545a172c9f77674525471ebeda7d6c86140e7a778f67ded92260.json')
+  const keyH = import('./fixtures/keys/identity-keys/03e0480538c2a39951d054e17ff31fde487cb1031d0044a037b53ad2e028a3e77c.json')
+
   // If in browser, put the fixture keys in local storage
   // so that Keystore can find them
-  const levelup = require('levelup')
-  const level = require('level-js')
   const signingStore = levelup(level('./orbitdb/identity/signingkeys'))
   const identityStore = levelup(level('./orbitdb/identity/identitykeys'))
```
[the remark this answers is not in the document]
Comment (haydenyoung) [reply test/browser.spec.js]: Could you please review https://github.com/orbitdb/ipfs-log/pull/363/commits/910cdbb2ef7df5beb3c34cbd411569b10b214d81?

Moving from levelup/leveldown (which are now deprecated) to abstract-level (I.e. new Level()) needs to be applied to other repos as well.

[the remark this answers is not in the document]
Comment (haadcode) [reply test/browser.spec.js]: https://github.com/orbitdb/ipfs-log/commit/910cdbb2ef7df5beb3c34cbd411569b10b214d81 looks good to me 👍

File: test/log-load.spec.js (hunk 9 above)
Comment (haydenyoung) [reply test/log-load.spec.js:881]: I notice that io is exported by utils/index.js. Would it make more sense to use the orbit-db-io module directly and destruct the functions?

E.g.

```
import { read, write } from 'orbit-db-io'
```

**Review by haadcode — COMMENTED** [review]
Let's also remove lib/es5 as discussed.

File: test/log-load.spec.js (hunk 9 above)
Comment (haydenyoung) [reply test/log-load.spec.js:881]: Have opted for destructing read, write from orbit-db-io. Seems more elegant and one less place to go looking for the export.

**Comment by haydenyoung** [comment]
> Let's also remove lib/es5 as discussed.

@haadcode should we add lib/es5 to gitignore to avoid being re-added if someone runs the build:es5 tools?

**Comment by haadcode** [comment]
> > Let's also remove lib/es5 as discussed.
> 
> @haadcode should we add lib/es5 to gitignore to avoid being re-added if someone runs the build:es5 tools?

Good idea 👍

**Review by haadcode — APPROVED** [review]

**Comment by haadcode** [comment]
LGTM 👍 

Awesome work @haydenyoung 🎉 🙏

**Pull request merged**